filters/auth/grantlogout.go

```diff
@@ -0,0 +1,202 @@
+//
+// grantLogout filter
+//
+// Used for revoking access/refresh tokens generated by the oauthGrant() filter.
+//
+
+package auth
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"strings"
+
+	log "github.com/sirupsen/logrus"
+	"github.com/zalando/skipper/filters"
+)
+
+const (
+	GrantLogoutName           = "grantLogout"
+	revokeTokenKey            = "token"
+	revokeTokenTypeKey        = "token_type_hint"
+	refreshTokenType          = "refresh_token"
+	accessTokenType           = "access_token"
+	unsupportedTokenTypeError = "unsupported_token_type"
+)
+
+type grantLogoutSpec struct {
+	config OAuthConfig
+}
+
+type grantLogoutFilter struct {
+	config OAuthConfig
+}
+
+type revokeErrorResponse struct {
+	Error            string `json:"error"`
+	ErrorDescription string `json:"error_description"`
+}
+
+type basicAuthCredentials struct {
+	clientId     string
+	clientSecret string
+}
+
+func (*grantLogoutSpec) Name() string { return GrantLogoutName }
+
+func (s *grantLogoutSpec) CreateFilter([]interface{}) (filters.Filter, error) {
+	return &grantLogoutFilter{
+		config: s.config,
+	}, nil
+}
+
+func (f *grantLogoutFilter) getBasicAuthCredentials() (*basicAuthCredentials, error) {
+	clientID := f.config.GetClientID()
+	if clientID == "" {
+		return nil, errors.New("failed to create token revoke auth header: no client ID")
+	}
+
+	clientSecret := f.config.GetClientSecret()
+	if clientSecret == "" {
+		return nil, errors.New("failed to create token revoke auth header: no client secret")
+	}
+
+	return &basicAuthCredentials{
+		clientId:     clientID,
+		clientSecret: clientSecret,
+	}, nil
+}
+
+func getErrorResponse(response *http.Response) *revokeErrorResponse {
+	buf, err := ioutil.ReadAll(response.Body)
+	if err == nil && buf != nil {
+		var errorResponse revokeErrorResponse
+		err = json.Unmarshal(buf, &errorResponse)
+
+		if err == nil {
+			return &errorResponse
+		}
+	}
+
+	return nil
+}
+
+func responseToError(response *http.Response, tokenType string) error {
+	errorResponse := getErrorResponse(response)
+	if errorResponse.Error == unsupportedTokenTypeError && tokenType == accessTokenType {
+		// Provider does not support revoking access tokens, which can happen according to RFC 7009.
+		// In that case this is not really an error.
+		return nil
+	}
+	return fmt.Errorf(
+		"%s revocation failed: %d %s",
+		tokenType,
+		response.StatusCode,
+		errorResponse.Error+": "+errorResponse.ErrorDescription,
```

Comment:
Format string and concatenation should not be mixed, should be changed to `"%s revocation failed: %d %s: %s"`

Reply:
Fixed.